src/main/java/com/walmartlabs/x12/rule/UniqueDocumentX12Rule.java

```diff
@@ -0,0 +1,59 @@
+package com.walmartlabs.x12.rule;
+
+import com.walmartlabs.x12.SegmentIterator;
+import com.walmartlabs.x12.X12Segment;
+import com.walmartlabs.x12.exceptions.X12ErrorDetail;
+import com.walmartlabs.x12.exceptions.X12ParserException;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * turns out that different documents
+ * have the document number and date in 
+ * different spots on the segment line
+ * 
+ * for example the document number
+ * on the ASN can be found at BSN02 
+ * but the document number 
+ * on the PO can be found at BEG03
+ * 
+ * therefore this rule can be configured
+ * to look for a specific identifier
+ * and will take two locations
+ * one for the document number and
+ * the other for the document date
+ */
+public class UniqueDocumentX12Rule implements X12Rule {
+
+    private final String segmentId;
+    private final int documentNumberIndex;
+    private final int documentDateIndex;
+    
+    public UniqueDocumentX12Rule(String segmentId, int documentNumberIndex, int documentDateIndex) {
+        this.segmentId = segmentId;
+        this.documentNumberIndex = documentNumberIndex;
+        this.documentDateIndex = documentDateIndex;
+    }
+    
+    @Override
+    public void verify(List<X12Segment> segmentList) {
+        Set<String> documentNumbers = new HashSet<>();
+
+        SegmentIterator segments = new SegmentIterator(segmentList);
+        while (segments.hasNext()) {
+            X12Segment currentSegment = segments.next();
+            if (segmentId.equals(currentSegment.getIdentifier())) {
+                String docNumber = currentSegment.getElement(documentNumberIndex);
+                String docDate = currentSegment.getElement(documentDateIndex);
+                if (!documentNumbers.add(docNumber + "_" + docDate)) {
```

Comment:
Should we do null check for `docNumber` & `docDate`?

Reply:
Should this be checked for transaction set type, document number & document date? 🤔

Reply:
Are they optional?

Reply:
will add some tests to demonstrate what happens when these values are missing, but not going to check for null/missing values. 

The reason for that, is that as a general principle: I really don't think that the X12Rule provided by Gozer (at least not those used by the splitter) should be doing much validation and evaluation beyond spot checking the whole EDI message (the main elements are in the correct order (ISA/IEA, GS/GE, and ST/SE) and things that would need to be done across transactions. Most of the other validation should be done on an individual document/transaction after it has been parsed. This allows one transaction to be rejected from the EDI message, while another can be processed. 

For example, I don't think by default the entire message should be rejected if a document number is missing on a single document/transaction. However, if two document/transactions are both missing document numbers that is flagged as an issue (there are two transactions that are missing the value). 

That said, consumers/users of the Gozer framework can easily create their own rules for different behaviors b/c the splitter and the rules were designed with extensibility in mind